lib/src/main/java/com/otaliastudios/transcoder/internal/thumbnails/DefaultThumbnailsEngine.kt

```diff
@@ -76,26 +78,68 @@ class DefaultThumbnailsEngine(
 
     private val stubs = ArrayDeque<Stub>()
 
+    private inner class IgnoringEosDataSource(
+        private val source: DataSource,
+    ) : DataSource by source {
+        override fun requestKeyFrameTimestamps(): Long {
+            return source.requestKeyFrameTimestamps()
+        }
+        override fun getKeyFrameTimestampsUs(): ArrayList<Long> {
+            return source.keyFrameTimestampsUs
+        }
+
+        override fun isDrained(): Boolean {
+            if(source.isDrained) {
+                source.seekTo(stubs.firstOrNull()?.positionUs ?: -1)
+            }
+            return source.isDrained
+        }
+    }
+    private fun DataSource.ignoringEOS(): DataSource = IgnoringEosDataSource(this)
+
     private fun createPipeline(
         type: TrackType,
         index: Int,
         status: TrackStatus,
         outputFormat: MediaFormat
     ): Pipeline {
-        val source = dataSources[type][index]
+        val source = dataSources[type][index].ignoringEOS()
+
         return Pipeline.build("Thumbnails") {
-            Seeker(source, fetchPosition) {
-                val nextSeek = (ceil(source.positionUs.toDouble() / KEY_FRAME) * KEY_FRAME).toLong()
-                val position = stubs.firstOrNull()?.positionUs ?: -1
+            Seeker(source) {
+                var seek = false
+                val seekUs: Long
+                val current = source.positionUs
+                val requested = stubs.firstOrNull()?.positionUs ?: -1
+                val threshold = stubs.firstOrNull()?.request?.threshold() ?: 0L
+                val nextKeyFrameIndex = source.search(requested)
```

Comment:
when requested == -1, it should early exit.
Basically, I feel we can move this code down

Reply:
check now @sahilbajaj